File: BOJ/Swift/1987.swift
let n = readLine()!.split(separator: " ").map { Int(String($0))! }
let row = n[0]
let col = n[1]
var adj: [[Int]] = []
for _ in 0..<row {
    let line = Array(readLine()!).map { Int($0.asciiValue! - 65) }
    adj.append(line)
}

let dx = [1, 0, -1, 0]
let dy = [0, 1, 0, -1]
var answer = 0

func dfs(x: Int, y: Int, total: Int, bitmask: Int) {
    answer = max(answer, total)
    for i in 0..<4 {
        let nx = x + dx[i]
        let ny = y + dy[i]

        if (nx < 0 || nx >= col || ny < 0 || ny >= row) {
            continue
        }
        let newBitmask = 1 << adj[ny][nx]
        if bitmask & newBitmask == 0 {
            dfs(x: nx, y: ny, total: total + 1, bitmask: bitmask | newBitmask)
        }
    }
}

dfs(x: 0, y: 0, total: 1, bitmask: 1 << adj[0][0])
print(answer)

File: BOJ/C++/1987.cpp
#include <iostream>
#include <vector>

using namespace std;

int row, col;
int adj[22][22];
string input;
bool alphabet[26];
bool visited[22][22];
int answer = 0;
int dx[4] = { 1, 0, -1, 0 };
int dy[4] = { 0, 1, 0, -1 };

void dfs(int x, int y, int total) {
	answer = max(answer, total);

	for (int i = 0; i < 4; i++) {
		int nx = x + dx[i];
		int ny = y + dy[i];

		if (nx < 0 || nx >= col || ny < 0 || ny >= row) {
			continue;
		}
		if (visited[ny][nx]) {
			continue;
		}
		if (alphabet[adj[ny][nx]]) {
			continue;
		}
		visited[ny][nx] = true;
		alphabet[adj[ny][nx]] = true;
		dfs(nx, ny, total + 1);
		alphabet[adj[ny][nx]] = false;
		visited[ny][nx] = false;
	}
}

int main() {
	cin >> row >> col;

	for (int i = 0; i < row; i++) {
		cin >> input;
		for (int j = 0; j < col; j++) {
			adj[i][j] = input[j] - 'A';
		}
	}

	visited[0][0] = true;
	alphabet[adj[0][0]] = true;
	dfs(0, 0, 1);
	cout << answer;
	return 0;
}